sw/device/lib/dif/dif_spi_host.c

```diff
@@ -0,0 +1,382 @@
+  bool active;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    active = bitfield_bit32_read(reg, SPI_HOST_STATUS_ACTIVE_BIT);
+  } while (active);
+
+  // Wait for the spi host fifos to drain.
+  uint32_t txqd, rxqd;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    txqd = bitfield_field32_read(reg, SPI_HOST_STATUS_TXQD_FIELD);
+    rxqd = bitfield_field32_read(reg, SPI_HOST_STATUS_RXQD_FIELD);
+  } while (txqd != 0 || rxqd != 0);
+
+  // Clear the software reset request bit.
+  mmio_region_write32(
+      spi_host->base_addr, SPI_HOST_CONTROL_REG_OFFSET,
+      bitfield_bit32_write(0, SPI_HOST_CONTROL_SW_RST_BIT, false));
+}
+
+static void spi_host_enable(const dif_spi_host_t *spi_host, bool enable) {
+  mmio_region_write32(
+      spi_host->base_addr, SPI_HOST_CONTROL_REG_OFFSET,
+      bitfield_bit32_write(0, SPI_HOST_CONTROL_SPIEN_BIT, enable));
+}
+
+dif_result_t dif_spi_host_configure(const dif_spi_host_t *spi_host,
+                                    dif_spi_host_config_t config) {
+  if (spi_host == NULL) {
+    return kDifBadArg;
+  }
+  if (config.peripheral_clock_freq_hz == 0 || config.spi_clock == 0) {
+    return kDifBadArg;
+  }
+
+  uint32_t divider = (config.peripheral_clock_freq_hz / config.spi_clock) - 1;
+  if (divider & ~SPI_HOST_CONFIGOPTS_CLKDIV_0_MASK) {
+    return kDifBadArg;
+  }
+
+  spi_host_reset(spi_host);
+  uint32_t reg = 0;
+  reg =
+      bitfield_field32_write(reg, SPI_HOST_CONFIGOPTS_CLKDIV_0_FIELD, divider);
+  reg = bitfield_field32_write(reg, SPI_HOST_CONFIGOPTS_CSNIDLE_0_FIELD,
+                               config.chip_select.idle);
+  reg = bitfield_field32_write(reg, SPI_HOST_CONFIGOPTS_CSNTRAIL_0_FIELD,
+                               config.chip_select.trail);
+  reg = bitfield_field32_write(reg, SPI_HOST_CONFIGOPTS_CSNLEAD_0_FIELD,
+                               config.chip_select.lead);
+  reg = bitfield_bit32_write(reg, SPI_HOST_CONFIGOPTS_FULLCYC_0_BIT,
+                             config.full_cycle);
+  reg = bitfield_bit32_write(reg, SPI_HOST_CONFIGOPTS_CPHA_0_BIT, config.cpha);
+  reg = bitfield_bit32_write(reg, SPI_HOST_CONFIGOPTS_CPOL_0_BIT, config.cpol);
+  mmio_region_write32(spi_host->base_addr, SPI_HOST_CONFIGOPTS_REG_OFFSET, reg);
+  spi_host_enable(spi_host, true);
+  return kDifOk;
+}
+
+static void wait_ready(const dif_spi_host_t *spi_host) {
+  bool ready;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    ready = bitfield_bit32_read(reg, SPI_HOST_STATUS_READY_BIT);
+  } while (!ready);
+}
+
+static void wait_tx_fifo(const dif_spi_host_t *spi_host) {
+  uint32_t txqd;
+  do {
+    uint32_t reg =
+        mmio_region_read32(spi_host->base_addr, SPI_HOST_STATUS_REG_OFFSET);
+    txqd = bitfield_field32_read(reg, SPI_HOST_STATUS_TXQD_FIELD);
+  } while (txqd == 0xFF);
```

Comment:
@martin-lueker What value should I be checking against here?  `SPI_HOST_PARAM_TX_DEPTH`? (or that value minus 1?).

Reply:
What is the goal of this routine? Are you waiting to see if the tx_fifo has drained completely?  If so then you should wait until txqd is 0.

Reply:
This routine is meant to wait until the TXFIFO can accept at least one write.

Reply:
Hi Chris, 
   In this case it would be `while (txqd == SPI_HOST_PARAM_TX_DEPTH)`